File: common/src/main/java/org/valkyrienskies/clockwork/content/forces/EncasedFanController.java
package org.valkyrienskies.clockwork.content.forces;

import org.valkyrienskies.clockwork.content.propulsion.fan.EncasedFanCreateData;
import org.valkyrienskies.clockwork.content.propulsion.fan.EncasedFanData;
import org.valkyrienskies.clockwork.content.propulsion.fan.EncasedFanUpdateData;
import it.unimi.dsi.fastutil.Pair;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import net.minecraft.util.Mth;
import org.jetbrains.annotations.NotNull;
import org.joml.Vector3d;
import org.joml.Vector3dc;
import org.valkyrienskies.core.api.ships.PhysShip;
import org.valkyrienskies.core.api.ships.ServerShip;
import org.valkyrienskies.core.impl.api.ShipForcesInducer;
import org.valkyrienskies.core.impl.game.ships.PhysShipImpl;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

public class EncasedFanController implements ShipForcesInducer  {

    final Int2ObjectOpenHashMap<EncasedFanData> fanData = new Int2ObjectOpenHashMap<>();

    private final ConcurrentHashMap<Integer, EncasedFanUpdateData> fanUpdateData = new ConcurrentHashMap<>();

    private final ConcurrentLinkedQueue<Pair<Integer, EncasedFanCreateData>> createdFans = new ConcurrentLinkedQueue<>();

    private final ConcurrentLinkedQueue<Integer> removedFans = new ConcurrentLinkedQueue<>();

    private int nextFanID = 0;

    public static EncasedFanController getOrCreate(ServerShip ship) {
        if (ship.getAttachment(EncasedFanController.class) == null) {
            ship.saveAttachment(EncasedFanController.class, new EncasedFanController());
        }
        return ship.getAttachment(EncasedFanController.class);
    }

    @Override
    public void applyForces(@NotNull PhysShip physShip) {
        while (!createdFans.isEmpty()) {
            final Pair<Integer, EncasedFanCreateData> createData = createdFans.remove();
            fanData.put(createData.left(), new EncasedFanData(
                    createData.right().fanPos(),
                    createData.right().fanDir(),
                    createData.right().fanSpeed()
            ));
        }
        while (!removedFans.isEmpty()) {
            fanData.remove((int) removedFans.remove());
        }

        fanUpdateData.forEach((id, data) -> {
            EncasedFanData physData = fanData.get(id);
            if (physData == null) {
                return;
            }
            physData.fanSpeed = data.fanSpeed();
        });

        fanUpdateData.clear();

        for (EncasedFanData physData : fanData.values()) {
            Vector3dc force = computeForce(physData, ((PhysShipImpl) physShip));
            Vector3dc fanVector = physData.fanPos.add(0.5,0.5,0.5, new Vector3d()).sub(physShip.getTransform().getPositionInShip());
            physShip.applyRotDependentForceToPos(force, fanVector);
        }
    }

    private Vector3dc computeForce(EncasedFanData physData, PhysShipImpl physShip) {
        double speed = physData.fanSpeed;

        Vector3d dir = new Vector3d(physData.fanDir).mul(Math.signum(speed));

        double providedForce = (Math.abs(speed) * 36.00875);
        double airPress = airPressure(physShip.getTransform().getPositionInWorld());
        Vector3dc fanPosRelCenterMass = physShip.getTransform().getShipToWorld().transformPosition(physData.fanPos.add(0.5,0.5,0.5, new Vector3d()), new Vector3d()).sub(physShip.getTransform().getPositionInWorld(), new Vector3d());
        Vector3dc worldVelAtFan = physShip.getPoseVel().getOmega().cross(fanPosRelCenterMass, new Vector3d()).add(physShip.getPoseVel().getVel(), new Vector3d());
        double exhaustVel = exhaustVelocity();


        double factor = 1.0 - Mth.clamp(dir.dot(worldVelAtFan) / exhaustVel, 0.0, 1.0);
        if (!Double.isFinite(factor)) {
            factor = 0;
        }

        providedForce = providedForce * airPress * factor;


        Vector3dc force = dir.mul(providedForce);

        return force;
    }

    private double airPressure(Vector3dc pos) {
        final double SEA_LEVEL = 64.0;
        final double WORLD_HEIGHT = 320.0;
        final double FALLOFF_POINT = 192.0;

        double offset = Math.exp(-(WORLD_HEIGHT-SEA_LEVEL)/FALLOFF_POINT);
        double height = pos.y();
        double airPress = (Math.exp(-(height-SEA_LEVEL)/FALLOFF_POINT)-offset)/(1.0-offset);
        if (Double.isFinite(airPress)) {
            return Mth.clamp(airPress, 0, 1);
        } else {
            return 0.0;
        }
    }

    private double exhaustVelocity() {
        return 44.074;
    }

    public int addEncasedFan(EncasedFanCreateData data) {
        int id = nextFanID++;
        createdFans.add(Pair.of(id, data));
        return id;
    }

    public void removeEncasedFan(int id) {
        removedFans.add(id);
    }

    public void updateEncasedFan(int id, EncasedFanUpdateData data) {
        fanUpdateData.put(id, data);
    }

}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/forces/EncasedFanController.kt
package org.valkyrienskies.clockwork.content.forces

import com.fasterxml.jackson.annotation.JsonAutoDetect
import it.unimi.dsi.fastutil.Pair
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import net.minecraft.util.Mth
import net.minecraft.world.entity.animal.horse.Horse
import org.joml.Vector3d
import org.joml.Vector3dc
import org.valkyrienskies.clockwork.content.propulsion.singleton.fan.EncasedFanCreateData
import org.valkyrienskies.clockwork.content.propulsion.singleton.fan.EncasedFanData
import org.valkyrienskies.clockwork.content.propulsion.singleton.fan.EncasedFanUpdateData
import org.valkyrienskies.core.api.ships.PhysShip
import org.valkyrienskies.core.api.ships.ServerShip
import org.valkyrienskies.core.api.ships.ShipForcesInducer
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.function.BiConsumer

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
class EncasedFanController(
    override val appliers: HashMap<Int, EncasedFanData> = HashMap(),
    override val applierUpdateData: ConcurrentLinkedQueue<kotlin.Pair<Int, EncasedFanUpdateData>> = ConcurrentLinkedQueue(),
    override val createdAppliers: ConcurrentLinkedQueue<kotlin.Pair<Int, EncasedFanCreateData>> = ConcurrentLinkedQueue(),
    override val removedAppliers: ConcurrentLinkedQueue<Int> = ConcurrentLinkedQueue(),
    override var nextApplierID: Int = 0
) : MultiInstanceForceApplier<EncasedFanUpdateData, EncasedFanData, EncasedFanCreateData> {

    override fun applyForces(physShip: PhysShip) {
        super.applyForces(physShip)

        for (physData in appliers.values) {
            val force = computeForce(physData, physShip)
            val fanVector: Vector3dc =
                Vector3d(physData.position).add(0.5, 0.5, 0.5, Vector3d()).sub(physShip.transform.positionInShip)
            physShip.applyRotDependentForceToPos(force, fanVector)
        }
    }

    private fun computeForce(
        physData: EncasedFanData,
        physShip: PhysShip
    ): Vector3dc {
        val speed: Double = physData.fanSpeed
        val dir: Vector3d = Vector3d(physData.fanDir).mul(Math.signum(speed))
        var providedForce = Math.abs(speed) * 36.00875
        val airPress = airPressure(physShip.transform.positionInWorld)
        val fanPosRelCenterMass: Vector3dc = physShip.transform.shipToWorld.transformPosition(
            Vector3d(physData.position).add(0.5, 0.5, 0.5, Vector3d()),
            Vector3d()
        ).sub(physShip.transform.positionInWorld, Vector3d())
        val worldVelAtFan: Vector3dc = physShip.omega.cross(fanPosRelCenterMass, Vector3d())
            .add(physShip.velocity, Vector3d())
        val exhaustVel = exhaustVelocity()
        var factor = 1.0 - Mth.clamp(dir.dot(worldVelAtFan) / exhaustVel, 0.0, 1.0)
        if (!java.lang.Double.isFinite(factor)) {
            factor = 0.0
        }

        providedForce *= airPress * factor
        return dir.mul(providedForce)
    }

    private fun airPressure(pos: Vector3dc): Double {
        val SEA_LEVEL = 64.0
        val WORLD_HEIGHT = 320.0
        val FALLOFF_POINT = 192.0
        val offset = Math.exp(-(WORLD_HEIGHT - SEA_LEVEL) / FALLOFF_POINT)
        val height = pos.y()
        val airPress = (Math.exp(-(height - SEA_LEVEL) / FALLOFF_POINT) - offset) / (1.0 - offset)
        return if (java.lang.Double.isFinite(airPress)) {
            Mth.clamp(airPress, 0.0, 1.0)
        } else {
            0.0
        }
    }

    private fun exhaustVelocity(): Double {
        return 44.074
    }

    companion object {
        fun getOrCreate(ship: ServerShip): EncasedFanController? {
            if (ship.getAttachment(EncasedFanController::class.java) == null) {
                ship.saveAttachment(EncasedFanController::class.java, EncasedFanController())
            }
            return ship.getAttachment(EncasedFanController::class.java)
        }
    }
}